### ISSUE TITLE
Unresponsive `lio_listio` when multiple `aiocb` structures are given

### ISSUE THREAD

**Issue by sandrain** [LLNL/UnifyFS#538]
### System information
Summitdev

### Describe the problem you're observing

`listio_lio` becomes unresponsive, when multiple `aiocb` is fed. I suspect that the resource manager encounters deadlock. Our existing example doesn't encounter this problem (possibly because only a single `aiocb` is passed at a time?).

### Describe how to reproduce the problem

Run the following example application, after creating `tt1` and `tt2` files (e.g., using `write-static` example):

```
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <errno.h>
#include <limits.h>
#include <sys/time.h>
#include <sys/types.h>
#include <aio.h>
#include <unistd.h>
#include <fcntl.h>
#include <libgen.h>
#include <assert.h>
#include <getopt.h>
#include <unifyfs.h>

static char buf[4*(1<<20)];

static int do_test_simple(void)
{
    int i;
    int ret;
    struct aiocb aio[2] = { 0, };
    struct aiocb *aio_list[2];
    int fd1, fd2;
    struct timeval t1, t2;
    double elapsed = 0.0;

    fprintf(stderr, "## == simple listio ==\n");

    fd1 = open("/unifyfs/tt1", O_RDONLY);
    fd2 = open("/unifyfs/tt2", O_RDONLY);
    assert(fd1 > 0 && fd2 > 0);

    for (i = 0; i < 2; i++) {
        int fd = i == 0 ? fd1 : fd2;
        unsigned long offset = i*(4*(1<<20));
        unsigned long count = 1<<20;
        struct aiocb *current = &aio[i];

        current->aio_fildes = fd;
        current->aio_offset = offset;
        current->aio_nbytes = count;
        current->aio_buf = &buf[i*(1<<20)];
        current->aio_lio_opcode = LIO_READ;

        aio_list[i] = current;

        fprintf(stderr, "## aio[%d] (fd=%d, offset=%lu, nbytes=%lu)\n",
                        i, current->aio_fildes,
                        current->aio_offset, current->aio_nbytes);
        fflush(stdout);
    }

    gettimeofday(&t1, NULL);
    fprintf(stderr, "## calling lio_listio..\n");

    ret = lio_listio(LIO_WAIT, aio_list, 2, 0);

    gettimeofday(&t2, NULL);
    elapsed = 1e-6*(t2.tv_sec*1e6 + t2.tv_usec - (t1.tv_sec*1e6 + t1.tv_usec));

    fprintf(stderr, "## lio_listio (took %.3lf sec): ret=%d, errno=%d (%s)\n",
                    elapsed, ret, errno, strerror(errno));

    for (i = 0; i < 2; i++) {
        ssize_t aio_ret = aio_return(&aio[i]);
        int error = aio_error(&aio[i]);
        fprintf(stderr, "## aio[%d] aio_ret = %zd (error=%d)\n",
                        i, aio_ret, error);
    }

    close(fd2);
    close(fd1);

    return ret;
}

static int do_test_mixed(void)
{
    int i;
    int ret;
    struct aiocb aio[4] = { 0, };
    struct aiocb *aio_list[4];
    int fd1, fd2;
    struct timeval t1, t2;
    double elapsed = 0.0;

    fprintf(stderr, "## == mixed listio ==\n");

    fd1 = open("/unifyfs/tt1", O_RDONLY);
    fd2 = open("/unifyfs/tt2", O_RDONLY);
    assert(fd1 > 0 && fd2 > 0);

    for (i = 0; i < 4; i++) {
        int fd = i < 2 ? fd1 : fd2;
        unsigned long offset = i*(4*(1<<20));
        unsigned long count = 1<<20;
        struct aiocb *current = &aio[i];

        current->aio_fildes = fd;
        current->aio_offset = offset;
        current->aio_nbytes = count;
        current->aio_buf = &buf[i*(1<<20)];
        current->aio_lio_opcode = LIO_READ;

        aio_list[i] = current;

        fprintf(stderr, "## aio[%d] (fd=%d, offset=%lu, nbytes=%lu)\n",
                        i, current->aio_fildes,
                        current->aio_offset, current->aio_nbytes);
        fflush(stdout);
    }

    gettimeofday(&t1, NULL);
    fprintf(stderr, "## calling lio_listio..\n");

    ret = lio_listio(LIO_WAIT, aio_list, 4, 0);

    gettimeofday(&t2, NULL);
    elapsed = 1e-6*(t2.tv_sec*1e6 + t2.tv_usec - (t1.tv_sec*1e6 + t1.tv_usec));

    fprintf(stderr, "## lio_listio (took %.3lf sec): ret=%d, errno=%d (%s)\n",
                    elapsed, ret, errno, strerror(errno));

    for (i = 0; i < 4; i++) {
        ssize_t aio_ret = aio_return(&aio[i]);
        int error = aio_error(&aio[i]);
        fprintf(stderr, "## aio[%d] aio_ret = %zd (error=%d)\n",
                        i, aio_ret, error);
    }

    close(fd2);
    close(fd1);

    return ret;
}

int main(int argc, char** argv)
{
    int ret = 0;

    ret = unifyfs_mount("/unifyfs", 0, 1, 0);
    if (ret) {
        fprintf(stderr, "unifyfs_mount failed (return = %d)\n", ret);
        return -1;
    }

    ret = do_test_simple();
    ret |= do_test_mixed();

    unifyfs_unmount();

    return ret;
}
```

This program runs successfully, without UnifyFS (using /tmp/ttX instead)

```
[root@rage17 tmp]$ dd if=/dev/zero of=/tmp/tt2 bs=16M count=10
[root@rage17 tmp]$ dd if=/dev/zero of=/tmp/tt2 bs=16M count=10
[root@rage17 tmp]$ ./read-listio
## == simple listio ==
## aio[0] (fd=3, offset=0, nbytes=1048576)
## aio[1] (fd=4, offset=4194304, nbytes=1048576)
## calling lio_listio..
## lio_listio (took 0.000 sec): ret=0, errno=0 (Success)
## aio[0] aio_ret = 1048576 (error=0)
## aio[1] aio_ret = 1048576 (error=0)
## == mixed listio ==
## aio[0] (fd=3, offset=0, nbytes=1048576)
## aio[1] (fd=3, offset=4194304, nbytes=1048576)
## aio[2] (fd=4, offset=8388608, nbytes=1048576)
## aio[3] (fd=4, offset=12582912, nbytes=1048576)
## calling lio_listio..
## lio_listio (took 0.000 sec): ret=0, errno=0 (Success)
## aio[0] aio_ret = 1048576 (error=0)
## aio[1] aio_ret = 1048576 (error=0)
## aio[2] aio_ret = 1048576 (error=0)
## aio[3] aio_ret = 1048576 (error=0)
```

### Include any warning or errors or relevant debugging data

This is the log message from the client:
```
2020-07-15T16:19:12 tid=88020 @ __wrap___xstat() [client/src/unifyfs-sysio.c:773] xstat was called for /tmp/unifyfs.hs2.427728/kvstore
2020-07-15T16:19:12 tid=88020 @ rpc_lookup_local_server_addr() [common/src/unifyfs_rpc_util.c:85] found local server rpc address 'na+sm://87690/0'
2020-07-15T16:19:12 tid=88020 @ unifyfs_client_rpc_init() [client/src/margo_client.c:66] svr_addr:'na+sm://87690/0' proto:'na+sm'
2020-07-15T16:19:12 tid=88020 @ unifyfs_mount() [client/src/unifyfs.c:2548] calling mount rpc
2020-07-15T16:19:12 tid=88020 @ invoke_client_mount_rpc() [client/src/margo_client.c:235] invoking the mount rpc function in client
2020-07-15T16:19:12 tid=88020 @ invoke_client_mount_rpc() [client/src/margo_client.c:248] Got response ret=0
2020-07-15T16:19:12 tid=88020 @ invoke_client_mount_rpc() [client/src/margo_client.c:257] My client id is 49
2020-07-15T16:19:12 tid=88020 @ unifyfs_init() [client/src/unifyfs.c:2302] FD limit for system = 4096
2020-07-15T16:19:12 tid=88020 @ init_superblock_structures() [client/src/unifyfs.c:2070] Meta-stacks initialized!
2020-07-15T16:19:12 tid=88020 @ unifyfs_logio_init_client() [common/src/unifyfs_logio.c:316] using spillover - size = 137438953472 B
2020-07-15T16:19:12 tid=88020 @ unifyfs_mount() [client/src/unifyfs.c:2565] calling attach rpc
2020-07-15T16:19:12 tid=88020 @ invoke_client_attach_rpc() [client/src/margo_client.c:194] invoking the attach rpc function in client
2020-07-15T16:19:12 tid=88020 @ invoke_client_attach_rpc() [client/src/margo_client.c:203] Got response ret=0
2020-07-15T16:19:12 tid=88020 @ invoke_client_metaget_rpc() [client/src/margo_client.c:368] invoking the metaget rpc function in client
2020-07-15T16:19:12 tid=88020 @ invoke_client_metaget_rpc() [client/src/margo_client.c:377] Got response ret=0
## == simple listio ==
2020-07-15T16:19:12 tid=88020 @ unifyfs_fid_open() [client/src/unifyfs.c:1764] unifyfs_get_fid_from_path() gave -1 (gfid = 717229742)
2020-07-15T16:19:12 tid=88020 @ invoke_client_metaget_rpc() [client/src/margo_client.c:368] invoking the metaget rpc function in client
2020-07-15T16:19:12 tid=88020 @ invoke_client_metaget_rpc() [client/src/margo_client.c:377] Got response ret=0
2020-07-15T16:19:12 tid=88020 @ unifyfs_fid_alloc() [client/src/unifyfs.c:1508] unifyfs_stack_pop() gave 0
2020-07-15T16:19:12 tid=88020 @ unifyfs_fid_create_file() [client/src/unifyfs.c:1552] Filename /unifyfs/tt1 got unifyfs fid 0
2020-07-15T16:19:12 tid=88020 @ unifyfs_fid_update_file_meta() [client/src/unifyfs.c:1393] laminated file size is 268435456 bytes
2020-07-15T16:19:12 tid=88020 @ unifyfs_fid_open() [client/src/unifyfs.c:1894] UNIFYFS_open generated fd 0 for file /unifyfs/tt1
2020-07-15T16:19:12 tid=88020 @ __wrap_open() [client/src/unifyfs-sysio.c:1126] UNIFYFS_open generated fd 0 for file /unifyfs/tt1
2020-07-15T16:19:12 tid=88020 @ unifyfs_fid_open() [client/src/unifyfs.c:1764] unifyfs_get_fid_from_path() gave -1 (gfid = 264536746)
2020-07-15T16:19:12 tid=88020 @ invoke_client_metaget_rpc() [client/src/margo_client.c:368] invoking the metaget rpc function in client
2020-07-15T16:19:12 tid=88020 @ invoke_client_metaget_rpc() [client/src/margo_client.c:377] Got response ret=0
2020-07-15T16:19:12 tid=88020 @ unifyfs_fid_alloc() [client/src/unifyfs.c:1508] unifyfs_stack_pop() gave 1
2020-07-15T16:19:12 tid=88020 @ unifyfs_fid_create_file() [client/src/unifyfs.c:1552] Filename /unifyfs/tt2 got unifyfs fid 1
2020-07-15T16:19:12 tid=88020 @ unifyfs_fid_update_file_meta() [client/src/unifyfs.c:1393] laminated file size is 268435456 bytes
2020-07-15T16:19:12 tid=88020 @ unifyfs_fid_open() [client/src/unifyfs.c:1894] UNIFYFS_open generated fd 1 for file /unifyfs/tt2
2020-07-15T16:19:12 tid=88020 @ __wrap_open() [client/src/unifyfs-sysio.c:1126] UNIFYFS_open generated fd 1 for file /unifyfs/tt2
## aio[0] (fd=4096, offset=0, nbytes=1048576)
## aio[1] (fd=4097, offset=4194304, nbytes=1048576)
## calling lio_listio..
2020-07-15T16:19:12 tid=88020 @ unifyfs_intercept_fd() [client/src/unifyfs.c:344] Changing fd from exposed 4096 to internal 0
2020-07-15T16:19:12 tid=88020 @ unifyfs_intercept_fd() [client/src/unifyfs.c:344] Changing fd from exposed 4097 to internal 1
2020-07-15T16:19:12 tid=88020 @ delegator_signal() [client/src/unifyfs.c:600] receive buffer now empty
2020-07-15T16:19:12 tid=88020 @ unifyfs_gfid_read_reqs() [client/src/unifyfs.c:1078] mread: n_reqs:2, flatcc buffer (0x1003af8f730) sz:70
2020-07-15T16:19:12 tid=88020 @ invoke_client_mread_rpc() [client/src/margo_client.c:639] invoking the read rpc function in client
```
, and server:
```
2020-07-15T16:19:12 tid=87692 @ unifyfs_attach_rpc() [server/src/unifyfs_cmd_handler.c:134] attaching client 1511587981:49
2020-07-15T16:19:12 tid=87692 @ rm_cmd_mread() [server/src/unifyfs_request_manager.c:1391] gfid:264536746, offset:4194304, length:1048576
2020-07-15T16:19:12 tid=87692 @ reserve_read_req() [server/src/unifyfs_request_manager.c:240] locking RM[1511587981:49] state
2020-07-15T16:19:12 tid=87692 @ reserve_read_req() [server/src/unifyfs_request_manager.c:257] reserved read req 0 (active=1, next=1)
2020-07-15T16:19:12 tid=87692 @ debug_print_read_req() [server/src/unifyfs_request_manager.c:233] server_read_req[0] status=0, gfid=0, num_remote=0
2020-07-15T16:19:12 tid=87692 @ reserve_read_req() [server/src/unifyfs_request_manager.c:262] unlocking RM[1511587981:49] state
2020-07-15T16:19:12 tid=87692 @ create_chunk_requests() [server/src/unifyfs_request_manager.c:348] locking RM[1511587981:49] state
2020-07-15T16:19:12 tid=87692 @ create_chunk_requests() [server/src/unifyfs_request_manager.c:350] creating chunk requests for rdreq 0
2020-07-15T16:19:12 tid=87692 @ debug_log_key_val() [server/src/unifyfs_metadata.c:67] @create_chunk_requests - key(gfid=264536746, offset=4194304), val(del=1, len=1048576, addr=4194304, app=1511587981, rank=8)
2020-07-15T16:19:12 tid=87692 @ create_chunk_requests() [server/src/unifyfs_request_manager.c:456] unlocking RM[1511587981:49] state
2020-07-15T16:19:12 tid=88023 @ rm_delegate_request_thread() [server/src/unifyfs_request_manager.c:2050] RM[1511587981:49] got work
2020-07-15T16:19:12 tid=88023 @ rm_request_remote_chunks() [server/src/unifyfs_request_manager.c:1674] read req 0 is active
2020-07-15T16:19:12 tid=88023 @ debug_print_read_req() [server/src/unifyfs_request_manager.c:233] server_read_req[0] status=1, gfid=264536746, num_remote=1
2020-07-15T16:19:12 tid=88023 @ rm_request_remote_chunks() [server/src/unifyfs_request_manager.c:1694] [0 of 1] sending 1 chunk requests to server 1
2020-07-15T16:19:12 tid=88023 @ sm_issue_chunk_reads() [server/src/unifyfs_service_manager.c:155] issuing 1 requests, total data size = 1048576
2020-07-15T16:19:12 tid=88023 @ sm_issue_chunk_reads() [server/src/unifyfs_service_manager.c:179] reading chunk(offset=4194304, size=1048576)
2020-07-15T16:19:12 tid=88023 @ sm_issue_chunk_reads() [server/src/unifyfs_service_manager.c:230] responding to myself
2020-07-15T16:19:12 tid=88023 @ rm_post_chunk_read_responses() [server/src/unifyfs_request_manager.c:1847] locking RM[1511587981:49] state
2020-07-15T16:19:12 tid=88023 @ rm_post_chunk_read_responses() [server/src/unifyfs_request_manager.c:1862] posting chunk responses for req 0 from delegator 1
2020-07-15T16:19:12 tid=88023 @ rm_post_chunk_read_responses() [server/src/unifyfs_request_manager.c:1878] unlocking RM[1511587981:49] state
2020-07-15T16:19:12 tid=88023 @ rm_delegate_request_thread() [server/src/unifyfs_request_manager.c:2065] unlocking RM[1511587981:49] state
2020-07-15T16:19:12 tid=88023 @ rm_delegate_request_thread() [server/src/unifyfs_request_manager.c:2020] locking RM[1511587981:49] state
2020-07-15T16:19:12 tid=88023 @ rm_process_remote_chunk_responses() [server/src/unifyfs_request_manager.c:1744] found read req 0 responses from delegator 1
2020-07-15T16:19:12 tid=88023 @ rm_handle_chunk_read_responses() [server/src/unifyfs_request_manager.c:1917] locking RM[1511587981:49] state
2020-07-15T16:19:12 tid=88023 @ rm_handle_chunk_read_responses() [server/src/unifyfs_request_manager.c:1932] handling chunk read responses from server 1: gfid=264536746 num_chunks=1 buf_size=1048600
2020-07-15T16:19:12 tid=88023 @ rm_handle_chunk_read_responses() [server/src/unifyfs_request_manager.c:1945] chunk response for offset=4194304: sz=1048576
2020-07-15T16:19:12 tid=88023 @ reserve_shmem_meta() [server/src/unifyfs_request_manager.c:1797] shm_data_header(cnt=0, bytes=0)
2020-07-15T16:19:12 tid=88023 @ reserve_shmem_meta() [server/src/unifyfs_request_manager.c:1819] reserved shm_data_meta[0] and 1048576 payload bytes
2020-07-15T16:19:12 tid=88023 @ client_signal() [server/src/unifyfs_request_manager.c:469] setting data-complete
2020-07-15T16:19:12 tid=87692 @ rm_cmd_mread() [server/src/unifyfs_request_manager.c:1391] gfid:717229742, offset:0, length:1048576
2020-07-15T16:19:12 tid=87692 @ reserve_read_req() [server/src/unifyfs_request_manager.c:240] locking RM[1511587981:49] state
```

It seems like the thread fails to acquire the mutex. I can probably debug more tomorrow.

**Comment by sandrain** [comment]
I think I found the bug (race condition) and fixed it. I will cleanup my fix and submit PR.

**Comment by sandrain** [comment]
After #539:

```
2020-07-17T10:08:18 tid=112685 @ __wrap___xstat() [client/src/unifyfs-sysio.c:773] xstat was called for /tmp/unifyfs.hs2.427836/kvstore
2020-07-17T10:08:18 tid=112685 @ rpc_lookup_local_server_addr() [common/src/unifyfs_rpc_util.c:85] found local server rpc address 'na+sm://112385/0'
2020-07-17T10:08:18 tid=112685 @ unifyfs_client_rpc_init() [client/src/margo_client.c:66] svr_addr:'na+sm://112385/0' proto:'na+sm'
2020-07-17T10:08:18 tid=112685 @ unifyfs_mount() [client/src/unifyfs.c:2601] calling mount rpc
2020-07-17T10:08:18 tid=112685 @ invoke_client_mount_rpc() [client/src/margo_client.c:235] invoking the mount rpc function in client
2020-07-17T10:08:18 tid=112685 @ invoke_client_mount_rpc() [client/src/margo_client.c:248] Got response ret=0
2020-07-17T10:08:18 tid=112685 @ invoke_client_mount_rpc() [client/src/margo_client.c:257] My client id is 49
2020-07-17T10:08:18 tid=112685 @ unifyfs_init() [client/src/unifyfs.c:2355] FD limit for system = 4096
2020-07-17T10:08:18 tid=112685 @ init_superblock_structures() [client/src/unifyfs.c:2123] Meta-stacks initialized!
2020-07-17T10:08:18 tid=112685 @ unifyfs_logio_init_client() [common/src/unifyfs_logio.c:316] using spillover - size = 137438953472 B
2020-07-17T10:08:18 tid=112685 @ unifyfs_mount() [client/src/unifyfs.c:2618] calling attach rpc
2020-07-17T10:08:18 tid=112685 @ invoke_client_attach_rpc() [client/src/margo_client.c:194] invoking the attach rpc function in client
2020-07-17T10:08:18 tid=112685 @ invoke_client_attach_rpc() [client/src/margo_client.c:203] Got response ret=0
2020-07-17T10:08:18 tid=112685 @ invoke_client_metaget_rpc() [client/src/margo_client.c:368] invoking the metaget rpc function in client
2020-07-17T10:08:18 tid=112685 @ invoke_client_metaget_rpc() [client/src/margo_client.c:377] Got response ret=0
## == simple listio ==
2020-07-17T10:08:18 tid=112685 @ unifyfs_fid_open() [client/src/unifyfs.c:1817] unifyfs_get_fid_from_path() gave -1 (gfid = 717229742)
2020-07-17T10:08:18 tid=112685 @ invoke_client_metaget_rpc() [client/src/margo_client.c:368] invoking the metaget rpc function in client
2020-07-17T10:08:18 tid=112685 @ invoke_client_metaget_rpc() [client/src/margo_client.c:377] Got response ret=0
2020-07-17T10:08:18 tid=112685 @ unifyfs_fid_alloc() [client/src/unifyfs.c:1561] unifyfs_stack_pop() gave 0
2020-07-17T10:08:18 tid=112685 @ unifyfs_fid_create_file() [client/src/unifyfs.c:1605] Filename /unifyfs/tt1 got unifyfs fid 0
2020-07-17T10:08:18 tid=112685 @ unifyfs_fid_update_file_meta() [client/src/unifyfs.c:1446] laminated file size is 68719476736 bytes
2020-07-17T10:08:18 tid=112685 @ unifyfs_fid_open() [client/src/unifyfs.c:1947] UNIFYFS_open generated fd 0 for file /unifyfs/tt1
2020-07-17T10:08:18 tid=112685 @ __wrap_open() [client/src/unifyfs-sysio.c:1126] UNIFYFS_open generated fd 0 for file /unifyfs/tt1
2020-07-17T10:08:18 tid=112685 @ unifyfs_fid_open() [client/src/unifyfs.c:1817] unifyfs_get_fid_from_path() gave -1 (gfid = 264536746)
2020-07-17T10:08:18 tid=112685 @ invoke_client_metaget_rpc() [client/src/margo_client.c:368] invoking the metaget rpc function in client
2020-07-17T10:08:18 tid=112685 @ invoke_client_metaget_rpc() [client/src/margo_client.c:377] Got response ret=0
2020-07-17T10:08:18 tid=112685 @ unifyfs_fid_alloc() [client/src/unifyfs.c:1561] unifyfs_stack_pop() gave 1
2020-07-17T10:08:18 tid=112685 @ unifyfs_fid_create_file() [client/src/unifyfs.c:1605] Filename /unifyfs/tt2 got unifyfs fid 1
2020-07-17T10:08:18 tid=112685 @ unifyfs_fid_update_file_meta() [client/src/unifyfs.c:1446] laminated file size is 68719476736 bytes
2020-07-17T10:08:18 tid=112685 @ unifyfs_fid_open() [client/src/unifyfs.c:1947] UNIFYFS_open generated fd 1 for file /unifyfs/tt2
2020-07-17T10:08:18 tid=112685 @ __wrap_open() [client/src/unifyfs-sysio.c:1126] UNIFYFS_open generated fd 1 for file /unifyfs/tt2
## aio[0] (fd=4096, offset=0, nbytes=1048576)
## aio[1] (fd=4097, offset=4194304, nbytes=1048576)
## calling lio_listio..
2020-07-17T10:08:18 tid=112685 @ unifyfs_intercept_fd() [client/src/unifyfs.c:344] Changing fd from exposed 4096 to internal 0
2020-07-17T10:08:18 tid=112685 @ unifyfs_intercept_fd() [client/src/unifyfs.c:344] Changing fd from exposed 4097 to internal 1
2020-07-17T10:08:18 tid=112685 @ delegator_signal() [client/src/unifyfs.c:600] receive buffer now empty
2020-07-17T10:08:18 tid=112685 @ unifyfs_gfid_read_reqs() [client/src/unifyfs.c:1118] mread: n_reqs:2, flatcc buffer (0x1001931f730) sz:70
2020-07-17T10:08:18 tid=112685 @ invoke_client_mread_rpc() [client/src/margo_client.c:639] invoking the read rpc function in client
2020-07-17T10:08:18 tid=112685 @ invoke_client_mread_rpc() [client/src/margo_client.c:648] Got response ret=0
2020-07-17T10:08:18 tid=112685 @ unifyfs_gfid_read_reqs() [client/src/unifyfs.c:1153] 2 read request(s) submitted to the server (2 files)
2020-07-17T10:08:18 tid=112685 @ process_read_data() [client/src/unifyfs.c:657] received 1 response from server
2020-07-17T10:08:18 tid=112685 @ process_read_data() [client/src/unifyfs.c:676] handling response[0] (gfid=264536746, offset=4194304, length=1048576, errcode=0)
2020-07-17T10:08:18 tid=112685 @ delegator_signal() [client/src/unifyfs.c:600] receive buffer now empty
2020-07-17T10:08:18 tid=112685 @ process_read_data() [client/src/unifyfs.c:657] received 1 response from server
2020-07-17T10:08:18 tid=112685 @ process_read_data() [client/src/unifyfs.c:676] handling response[0] (gfid=717229742, offset=0, length=1048576, errcode=0)
2020-07-17T10:08:18 tid=112685 @ delegator_signal() [client/src/unifyfs.c:600] receive buffer now empty
2020-07-17T10:08:18 tid=112685 @ unifyfs_gfid_read_reqs() [client/src/unifyfs.c:1181] fetched all data from server for 2 requests
2020-07-17T10:08:18 tid=112685 @ unifyfs_gfid_read_reqs() [client/src/unifyfs.c:1191] req[0] (gfid=264536746, offset=4194304, length=1048576, errcode=0)
2020-07-17T10:08:18 tid=112685 @ unifyfs_gfid_read_reqs() [client/src/unifyfs.c:1191] req[1] (gfid=717229742, offset=0, length=1048576, errcode=0)
2020-07-17T10:08:18 tid=112685 @ __wrap_lio_listio() [client/src/unifyfs-sysio.c:1628] finished handling 2 requests
## lio_listio (took 0.003 sec): ret=0, errno=0 (Success)
## aio[0] aio_ret = 1048576 (error=0)
## aio[1] aio_ret = 1048576 (error=0)
2020-07-17T10:08:18 tid=112685 @ unifyfs_intercept_fd() [client/src/unifyfs.c:344] Changing fd from exposed 4097 to internal 1
2020-07-17T10:08:18 tid=112685 @ __wrap_close() [client/src/unifyfs-sysio.c:2060] closing fd 1
2020-07-17T10:08:18 tid=112685 @ unifyfs_intercept_fd() [client/src/unifyfs.c:344] Changing fd from exposed 4096 to internal 0
2020-07-17T10:08:18 tid=112685 @ __wrap_close() [client/src/unifyfs-sysio.c:2060] closing fd 0
## == mixed listio ==
2020-07-17T10:08:18 tid=112685 @ unifyfs_get_fid_from_path() [client/src/unifyfs.c:399] File found: unifyfs_filelist[0].filename = /unifyfs/tt1
2020-07-17T10:08:18 tid=112685 @ unifyfs_fid_open() [client/src/unifyfs.c:1817] unifyfs_get_fid_from_path() gave 0 (gfid = 717229742)
2020-07-17T10:08:18 tid=112685 @ invoke_client_metaget_rpc() [client/src/margo_client.c:368] invoking the metaget rpc function in client
2020-07-17T10:08:18 tid=112685 @ invoke_client_metaget_rpc() [client/src/margo_client.c:377] Got response ret=0
2020-07-17T10:08:18 tid=112685 @ unifyfs_fid_update_file_meta() [client/src/unifyfs.c:1446] laminated file size is 68719476736 bytes
2020-07-17T10:08:18 tid=112685 @ unifyfs_fid_open() [client/src/unifyfs.c:1947] UNIFYFS_open generated fd 0 for file /unifyfs/tt1
2020-07-17T10:08:18 tid=112685 @ __wrap_open() [client/src/unifyfs-sysio.c:1126] UNIFYFS_open generated fd 0 for file /unifyfs/tt1
2020-07-17T10:08:18 tid=112685 @ unifyfs_get_fid_from_path() [client/src/unifyfs.c:399] File found: unifyfs_filelist[1].filename = /unifyfs/tt2
2020-07-17T10:08:18 tid=112685 @ unifyfs_fid_open() [client/src/unifyfs.c:1817] unifyfs_get_fid_from_path() gave 1 (gfid = 264536746)
2020-07-17T10:08:18 tid=112685 @ invoke_client_metaget_rpc() [client/src/margo_client.c:368] invoking the metaget rpc function in client
2020-07-17T10:08:18 tid=112685 @ invoke_client_metaget_rpc() [client/src/margo_client.c:377] Got response ret=0
2020-07-17T10:08:18 tid=112685 @ unifyfs_fid_update_file_meta() [client/src/unifyfs.c:1446] laminated file size is 68719476736 bytes
2020-07-17T10:08:18 tid=112685 @ unifyfs_fid_open() [client/src/unifyfs.c:1947] UNIFYFS_open generated fd 1 for file /unifyfs/tt2
2020-07-17T10:08:18 tid=112685 @ __wrap_open() [client/src/unifyfs-sysio.c:1126] UNIFYFS_open generated fd 1 for file /unifyfs/tt2
## aio[0] (fd=4096, offset=0, nbytes=1048576)
## aio[1] (fd=4096, offset=4194304, nbytes=1048576)
## aio[2] (fd=4097, offset=8388608, nbytes=1048576)
## aio[3] (fd=4097, offset=12582912, nbytes=1048576)
## calling lio_listio..
2020-07-17T10:08:18 tid=112685 @ unifyfs_intercept_fd() [client/src/unifyfs.c:344] Changing fd from exposed 4096 to internal 0
2020-07-17T10:08:18 tid=112685 @ unifyfs_intercept_fd() [client/src/unifyfs.c:344] Changing fd from exposed 4096 to internal 0
2020-07-17T10:08:18 tid=112685 @ unifyfs_intercept_fd() [client/src/unifyfs.c:344] Changing fd from exposed 4097 to internal 1
2020-07-17T10:08:18 tid=112685 @ unifyfs_intercept_fd() [client/src/unifyfs.c:344] Changing fd from exposed 4097 to internal 1
2020-07-17T10:08:18 tid=112685 @ delegator_signal() [client/src/unifyfs.c:600] receive buffer now empty
2020-07-17T10:08:18 tid=112685 @ unifyfs_gfid_read_reqs() [client/src/unifyfs.c:1118] mread: n_reqs:4, flatcc buffer (0x1001931f6a0) sz:118
2020-07-17T10:08:18 tid=112685 @ invoke_client_mread_rpc() [client/src/margo_client.c:639] invoking the read rpc function in client
2020-07-17T10:08:18 tid=112685 @ invoke_client_mread_rpc() [client/src/margo_client.c:648] Got response ret=0
2020-07-17T10:08:18 tid=112685 @ unifyfs_gfid_read_reqs() [client/src/unifyfs.c:1153] 4 read request(s) submitted to the server (2 files)
2020-07-17T10:08:18 tid=112685 @ process_read_data() [client/src/unifyfs.c:657] received 2 response from server
2020-07-17T10:08:18 tid=112685 @ process_read_data() [client/src/unifyfs.c:676] handling response[0] (gfid=264536746, offset=8388608, length=1048576, errcode=0)
2020-07-17T10:08:18 tid=112685 @ process_read_data() [client/src/unifyfs.c:676] handling response[1] (gfid=264536746, offset=12582912, length=1048576, errcode=0)
2020-07-17T10:08:18 tid=112685 @ delegator_signal() [client/src/unifyfs.c:600] receive buffer now empty
2020-07-17T10:08:18 tid=112685 @ process_read_data() [client/src/unifyfs.c:657] received 2 response from server
2020-07-17T10:08:18 tid=112685 @ process_read_data() [client/src/unifyfs.c:676] handling response[0] (gfid=717229742, offset=0, length=1048576, errcode=0)
2020-07-17T10:08:18 tid=112685 @ process_read_data() [client/src/unifyfs.c:676] handling response[1] (gfid=717229742, offset=4194304, length=1048576, errcode=0)
2020-07-17T10:08:18 tid=112685 @ delegator_signal() [client/src/unifyfs.c:600] receive buffer now empty
2020-07-17T10:08:18 tid=112685 @ unifyfs_gfid_read_reqs() [client/src/unifyfs.c:1181] fetched all data from server for 4 requests
2020-07-17T10:08:18 tid=112685 @ unifyfs_gfid_read_reqs() [client/src/unifyfs.c:1191] req[0] (gfid=264536746, offset=8388608, length=1048576, errcode=0)
2020-07-17T10:08:18 tid=112685 @ unifyfs_gfid_read_reqs() [client/src/unifyfs.c:1191] req[1] (gfid=264536746, offset=12582912, length=1048576, errcode=0)
2020-07-17T10:08:18 tid=112685 @ unifyfs_gfid_read_reqs() [client/src/unifyfs.c:1191] req[2] (gfid=717229742, offset=0, length=1048576, errcode=0)
2020-07-17T10:08:18 tid=112685 @ unifyfs_gfid_read_reqs() [client/src/unifyfs.c:1191] req[3] (gfid=717229742, offset=4194304, length=1048576, errcode=0)
2020-07-17T10:08:18 tid=112685 @ __wrap_lio_listio() [client/src/unifyfs-sysio.c:1628] finished handling 4 requests
## lio_listio (took 0.004 sec): ret=0, errno=0 (Success)
## aio[0] aio_ret = 1048576 (error=0)
## aio[1] aio_ret = 1048576 (error=0)
## aio[2] aio_ret = 1048576 (error=0)
## aio[3] aio_ret = 1048576 (error=0)
2020-07-17T10:08:18 tid=112685 @ unifyfs_intercept_fd() [client/src/unifyfs.c:344] Changing fd from exposed 4097 to internal 1
2020-07-17T10:08:18 tid=112685 @ __wrap_close() [client/src/unifyfs-sysio.c:2060] closing fd 1
2020-07-17T10:08:18 tid=112685 @ unifyfs_intercept_fd() [client/src/unifyfs.c:344] Changing fd from exposed 4096 to internal 0
2020-07-17T10:08:18 tid=112685 @ __wrap_close() [client/src/unifyfs-sysio.c:2060] closing fd 0
2020-07-17T10:08:18 tid=112685 @ unifyfs_unmount() [client/src/unifyfs.c:2669] syncing data
2020-07-17T10:08:18 tid=112685 @ unifyfs_unmount() [client/src/unifyfs.c:2681] calling unmount
2020-07-17T10:08:18 tid=112685 @ invoke_client_unmount_rpc() [client/src/margo_client.c:282] invoking the unmount rpc function in client
2020-07-17T10:08:18 tid=112685 @ invoke_client_unmount_rpc() [client/src/margo_client.c:291] Got response ret=0
```

**Comment by sandrain** [comment]
This was addressed by #539.
